ui/page/dexclient/add_dex_modal.go

```diff
@@ -180,49 +278,117 @@ func (md *addDexModal) Layout(gtx layout.Context) D {
+		childrens = append(childrens, layout.Rigid(func(gtx C) D {
+			return md.knownServers[host].Layout(gtx, func(gtx C) D {
+				gtx.Constraints.Min.X = gtx.Constraints.Max.X
+				return layout.Inset{
+					Top:    values.MarginPadding8,
+					Bottom: values.MarginPadding8,
+					Left:   values.MarginPadding12,
+					Right:  values.MarginPadding12,
+				}.Layout(gtx, func(gtx C) D {
+					gtx.Constraints.Min.Y = 45
+					return layout.Flex{Spacing: layout.SpaceBetween, Alignment: layout.Middle}.Layout(gtx,
+						layout.Rigid(md.Theme.Label(values.MarginPadding14, host).Layout),
+						layout.Rigid(func(gtx C) D {
+							if md.selectedServer != host {
+								return D{}
+							}
+							ic := decredmaterial.NewIcon(md.Theme.Icons.NavigationCheck)
+							ic.Color = md.Theme.Color.Success
+							return ic.Layout(gtx, values.MarginPadding20)
+						}),
+					)
+				})
+			})
+		}))
 	}
-	return fmt.Sprintf("%s\n\nThis DEX supports the following markets. All trades are in multiples of each market's lot size.\n\n%s", txt, strings.Join(markets, "\n"))
+
+	return layout.Flex{Axis: layout.Vertical}.Layout(gtx, childrens...)
+}
+
+func (md *addDexModal) customServerLayout(gtx C) D {
+	return layout.Flex{Axis: layout.Vertical}.Layout(gtx,
+		layout.Rigid(md.dexServerAddress.Layout),
+		layout.Rigid(func(gtx C) D {
+			gtx.Constraints.Max.Y = 300
+			return layout.Inset{Top: values.MarginPadding10}.Layout(gtx, md.cert.Layout)
+		}),
+	)
 }
 
-func formatAmount(assetID uint32, assetName string, amount uint64) string {
-	assetInfo, err := asset.Info(assetID)
-	if err != nil {
-		return fmt.Sprintf("%d [%s units]", amount, assetName)
+func payFeeAndRegister(l *load.Load, dexServer *core.Exchange, cert []byte, appPass string, onDexAdded func()) {
+	// Create the assetSelectorModal now, it'll remain open/visible
+	// until the fee is paid and registration is completed or the
+	// user manually closes it.
+	assetSelectorModal := newFeeAssetSelectorModal(l, dexServer)
+
+	confirmAndRegister := func(feeAsset *core.SupportedAsset, selectFeeAsset func()) {
+		modal.NewInfoModal(l).
+			Title(strConfirmReg).
+			Body(confirmRegisterModalDesc(dexServer, feeAsset.Symbol)).
+			SetCancelable(false).
+			NegativeButton(values.String(values.StrCancel), selectFeeAsset).
+			PositiveButton(strRegister, func(_ bool) {
+				selectFeeAsset()
+				go func() {
+					assetSelectorModal.modal.SetDisabled(true) // prevent re-selecting a fee asset
+					regFeeAsset := dexServer.RegFees[feeAsset.Symbol]
+					_, err := l.Dexc().RegisterWithDEXServer(dexServer.Host,
+						cert,
+						int64(regFeeAsset.Amt),
+						int32(regFeeAsset.ID),
+						[]byte(appPass))
+					if err != nil {
+						assetSelectorModal.modal.SetDisabled(false) // re-enable fee asset selection
+						assetSelectorModal.Toast.NotifyError(err.Error())
+						return
+					}
+					assetSelectorModal.Dismiss()
+					onDexAdded()
+				}()
+			}).Show()
+	}
+
+	var selectFeeAsset func()
+	selectFeeAsset = func() {
```

Comment:
Don't need a `selectFeeAsset` to re-display the fee asset selector modal, can just use `assetSelectorModal.Show` everywhere you currently have `selectFeeAsset`.